staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -137,6 +146,147 @@ type DiscoveryClient struct {
+	groups := []*metav1.APIGroup{}
+	if d.groups != nil {
+		for _, group := range d.groups.Groups {
+			groups = append(groups, &group)
+		}
+	}
+	return groups
+}
+
+// Returns all the stored discovery resources; empty slice if
+// the resources haven't been cached yet.
+func (d *DiscoveryClient) allResources() []*metav1.APIResourceList {
+	resources := []*metav1.APIResourceList{}
+	if d.resourcesByGV != nil {
+		for _, resourceList := range d.resourcesByGV {
+			resources = append(resources, resourceList)
+		}
+	}
+	return resources
+}
+
+// Returns all the stored "preferred" resources (highest priority
+// version); empty slice if resources haven't been cached yet.
+func (d *DiscoveryClient) preferredResources() []*metav1.APIResourceList {
+	resources := []*metav1.APIResourceList{}
+	if d.groups == nil || d.resourcesByGV == nil {
+		return resources
+	}
+	for _, group := range d.groups.Groups {
+		gv := group.PreferredVersion.GroupVersion
+		if resourceList, found := d.resourcesByGV[gv]; found {
+			resources = append(resources, resourceList)
+		}
+	}
+	return resources
+}
+
+// Returns all the stored "preferred" resources (highest priority
+// version); empty slice if resources haven't been cached yet.
+func (d *DiscoveryClient) preferredNamespacedResources() []*metav1.APIResourceList {
+	filteredResources := []*metav1.APIResourceList{}
+	for _, resourceList := range d.preferredResources() {
+		filteredResourceList := metav1.APIResourceList{}
+		for _, resource := range resourceList.APIResources {
+			if resource.Namespaced {
+				filteredResourceList.APIResources = append(filteredResourceList.APIResources, resource)
+			}
+		}
+		filteredResources = append(filteredResources, &filteredResourceList)
+	}
+	return filteredResources
+}
+
+// Stores the discovery groups. If "aggregated" discovery, resources stored also.
+// TODO(seans3): Need to figure out the mutex strategy here (if needed).
+func (d *DiscoveryClient) cacheDiscovery() error {
+	if d.isCached() {
+		return nil
+	}
+	gdl := apidiscoveryv2beta1.APIGroupDiscoveryList{}
+	body, err := d.restClient.Get().AbsPath("/apis").SetHeader("Accept", acceptV2Beta1).Do(context.TODO()).Raw()
+	// TODO(seans3): Is the "not acceptable" error how content negotiation works? I have not
+	// seen a way to get the http response content-type header (or any response header).
+	if err != nil && errors.IsNotAcceptable(err) {
+		return d.retrieveV1Discovery()
+	}
+	// to be compatible with a v1.0 server, if it's a 403 or 404 -- ignore.
+	// TODO(seans3): Do we still need this check to be compatible with v1.0?
+	if err != nil && !errors.IsNotFound(err) && !errors.IsForbidden(err) {
+		return err
+	}
+	if err = json.Unmarshal(body, &gdl); err != nil {
+		return err
+	}
+	groups, resources := parseDiscoveryV2Beta1(gdl)
+	d.lock.Lock()
+	d.version = versionV2Beta1
+	d.groups = groups
+	d.resourcesByGV = resources
+	d.lock.Unlock()
+	return nil
+}
+
+func (d *DiscoveryClient) retrieveV1Discovery() error {
```

Comment:
see other comment. I dont think the v1 and v2 discovery documents need to be retrieved in separate HTTP calls.